File: Javascript/n_queens.js
/**
 * @param {number} n
 * @return {string[][]}
 */
var SolveNqueensHelper = (board, n, col, asc_dig, dec_dig, queens_pos, result) => {
    var cur_row = queens_pos.length
    for (var cur_col = 0; cur_col < n; ++cur_col) {
        var check = false
        if (!col.has(cur_col) && !asc_dig.has(cur_row + cur_col) && !dec_dig.has(cur_row - cur_col)) {
            check = true
            queens_pos.push([cur_row, cur_col])
            col.add(cur_col)
            asc_dig.add(cur_row + cur_col)
            dec_dig.add(cur_row - cur_col)
            board[cur_row][cur_col] = 'Q'
            if (queens_pos.length === n) {
                var ret = []
                for (i in board) {
                    ret.push(board[i].join(''))
                }
                result.push(ret)
            }
            else {
                SolveNqueensHelper(board, n, col, asc_dig, dec_dig, queens_pos, result)
            }
        }
        if (check) {
            var top = queens_pos.pop()
            col.delete(top[1])
            asc_dig.delete(top[0] + top[1])
            dec_dig.delete(top[0] - top[1])
            board[top[0]][top[1]] = '.'
        }
    }
}

// Time complexity: O(nˆn)
// Space complexity: O(n)
var solveNQueens = function(n) {
    var board = [...Array(n)].map(x=>Array(n).fill('.'))
    var col = new Set()
    var asc_dig = new Set()
    var dec_dig = new Set()
    var result = []
    SolveNqueensHelper(board, n, col, asc_dig, dec_dig, [], result)
    return result
};

File: C++/n_queens.cpp
class Solution {
public:    
    void solveNQueensHelper(vector<string> board, int n, set<int> col, set<int> asc_diag, set<int> dec_diag, int start_row, vector<vector<string>>& result) {
        if (start_row == n) {
            result.push_back(board);
        }
        for (int cur_col = 0; cur_col < n; ++cur_col) {
            if (col.find(cur_col) == col.end() && asc_diag.find(start_row + cur_col) == asc_diag.end() && dec_diag.find(start_row - cur_col) == dec_diag.end()) {
                col.emplace(cur_col);
                asc_diag.emplace(start_row + cur_col);
                dec_diag.emplace(start_row - cur_col);
                board[start_row][cur_col] = 'Q';

                solveNQueensHelper(board, n, col, asc_diag, dec_diag, start_row + 1, result);

                col.erase(cur_col);
                asc_diag.erase(start_row + cur_col);
                dec_diag.erase(start_row - cur_col);
                board[start_row][cur_col] = '.';
            }
        }
    }
    
    vector<vector<string>> solveNQueens(int n) {
        string s(n,'.');
        vector<string> board(n, s);
        set<int> col;
        set<int> asc_diag;
        set<int> dec_diag;
        vector<vector<string>> result;
        solveNQueensHelper(board, n, col, asc_diag, dec_diag, 0, result);
        return result;
    }
};
